graph/push.go

```diff
@@ -287,63 +312,66 @@ func (s *TagStore) pushV2Repository(r *registry.Session, eng *engine.Engine, out
 		if err != nil {
 			return err
 		}
-		log.Infof("Signed manifest using daemon's key: %s", s.trustKey.KeyID())
-
-		manifestBytes = string(signedBody)
-	}
-
-	manifest, verified, err := s.verifyManifest(eng, []byte(manifestBytes))
-	if err != nil {
-		return fmt.Errorf("error verifying manifest: %s", err)
-	}
-
-	if err := checkValidManifest(manifest); err != nil {
-		return fmt.Errorf("invalid manifest: %s", err)
-	}
-
-	if !verified {
-		log.Debugf("Pushing unverified image")
-	}
+		log.Infof("Signed manifest for %s:%s using daemon's key: %s", repoInfo.LocalName, tag, s.trustKey.KeyID())
 
-	for i := len(manifest.FSLayers) - 1; i >= 0; i-- {
-		var (
-			sumStr  = manifest.FSLayers[i].BlobSum
-			imgJSON = []byte(manifest.History[i].V1Compatibility)
-		)
+		manifestBytes := string(signedBody)
 
-		sumParts := strings.SplitN(sumStr, ":", 2)
-		if len(sumParts) < 2 {
-			return fmt.Errorf("Invalid checksum: %s", sumStr)
+		manifest, verified, err := s.verifyManifest(eng, signedBody)
+		if err != nil {
+			return fmt.Errorf("error verifying manifest: %s", err)
 		}
-		manifestSum := sumParts[1]
 
-		img, err := image.NewImgJSON(imgJSON)
-		if err != nil {
-			return fmt.Errorf("Failed to parse json: %s", err)
+		if err := checkValidManifest(manifest); err != nil {
```

Comment:
Yeah, it is probably better to check that before, makes the code easier to follow.  In this case checkValidManifest should always return true since it was built locally, but I will keep it there for clarity that this check needs to be done when it is given from the client.